test/e2e-go/features/incentives/whalejoin_test.go

```diff
@@ -0,0 +1,324 @@
+// Copyright (C) 2019-2024 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package suspension
+
+import (
+	"fmt"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	v2 "github.com/algorand/go-algorand/daemon/algod/api/server/v2"
+	"github.com/algorand/go-algorand/daemon/algod/api/server/v2/generated/model"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/libgoal"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/test/framework/fixtures"
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+// TestWhaleJoin shows a "whale" with more stake than is currently online can go
+// online without immediate suspension.  This tests for a bug we had where we
+// calcululated expected proposal interval using the _old_ totals, rather than
+// the totals following the keyreg. So big joiner was being expected to propose
+// in the same block it joined.
+func TestWhaleJoin(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	defer fixtures.ShutdownSynchronizedTest(t)
+
+	t.Parallel()
+	a := require.New(fixtures.SynchronizedTest(t))
+
+	var fixture fixtures.RestClientFixture
+	// Make rounds shorter and seed lookback smaller, otherwise we need to wait
+	// 320 slow rounds for particpation effects to matter.
+	const lookback = 32
+	fixture.FasterConsensus(protocol.ConsensusFuture, time.Second, lookback)
+	fixture.Setup(t, filepath.Join("nettemplates", "Payouts.json"))
+	defer fixture.Shutdown()
+
+	// Overview of this test:
+	// 1. Take wallet15 offline (but retain keys so can back online later)
+	// 2. Have wallet01 spend almost all their algos
+	// 3. Wait for balances to flow through "lookback"
```

Comment:
What happens if wallet15 rejoins earlier, then wallet01 spends almost all of their algos? I guess that should not be subject to this test, but the idea being they become a whale while incentive eligible.

Reply:
That's much safer, because you would expect wallet15 to propose while wallet01 is spending.

I suppose there is an issue, similar to why we adding the "fee heartbeat when balance doubles", but it would only happen in very weird (maybe in tests) scenarios.  Only two wallets online,  1 algo and 100 algos.  If the 100 algos account spends down to 1 algo in one transaction, the original 1 algo account will go from 1% to 50%. So, if it has not proposed in 40 rounds (50% gives interval of 2, times 20 gives 40) it could get immediately knocked off.  This seems like something we don't need to cover.